### PR TITLE
Add loggers into IFrame host

### LINKED ISSUE #1421: inject logging interface into iframe socket storage drivers and validate errors are properly logger
https://github.com/microsoft/FluidFramework/pull/1328#discussion_r387825068

we should review all catches in iframe socket storage project and make sure we are logging and handling appropriately

### PR DESCRIPTION
**Pull request by heliocliu**
Fixes #1421 

Two goals here:
- Provide a way for the inner-IFrame driver to get the outer driver to log messages
- Provide a way for the inner-IFrame driver to log messages itself

The `ICombinedDriver` proxy object is currently functioning as the driver's internal props, so I added a minimal (`ITelemetryBaseLogger`) logger in there.  This lets the inner-IFrame driver to call `send` on that logger instead of needing to create its own.  Also update the `IDocumentService` and `IDocumentServiceFactory` implementations to more closely match the live ones wrt logger arguments, and add an example where the inner-IFrame driver creates and uses its own `ITelemetryBaseLogger` instead of using the proxy logger.

### CODE REVIEW

File: packages/drivers/iframe-driver/src/outerDocumentServiceFactory.ts [hunk 1]
```diff
@@ -69,9 +76,10 @@ export class DocumentServiceFactoryProxy implements IDocumentServiceFactoryProxy
         this.createProxy(frame);
     }
 
-    public async createDocumentService(): Promise<string> {
+    public async createDocumentService(logger?: ITelemetryBaseLogger): Promise<string> {
+        const outerProxyLogger = ChildLogger.create(logger, "OuterProxyIFrameDriver");
         const connectedDocumentService: IDocumentService =
-            await this.documentServiceFactory.createDocumentService(this.resolvedUrl);
+            await this.documentServiceFactory.createDocumentService(this.resolvedUrl, outerProxyLogger);
```
Comment (heliocliu) [inline packages/drivers/iframe-driver/src/outerDocumentServiceFactory.ts:82]: This part causes the proxy situation to be a bit different from the live one - the created driver creates a (private) child logger from the provided so it's not exactly the same as the one provided through the proxy

File: packages/drivers/iframe-driver/src/innerDocumentServiceFactory.ts [hunk 2]
```diff
@@ -29,6 +29,7 @@ export class InnerDocumentServiceFactory implements IDocumentServiceFactory {
                 outerProxy = Comlink.wrap<IDocumentServiceFactoryProxy>(Comlink.windowEndpoint(window.parent));
             }
             if (outerProxy) {
+                console.log(outerProxy.connected.toString());
```
Comment (ChumpChief) [inline packages/drivers/iframe-driver/src/innerDocumentServiceFactory.ts:32]: Remove?

File: packages/drivers/iframe-driver/src/innerDocumentService.ts [hunk 3]
```diff
@@ -22,16 +25,32 @@ export class InnerDocumentService implements IDocumentService {
...
+        this.logger = logger
+            ? ChildLogger.create(logger, "InnerIFrameDriver")
+            : new IFrameDebugLogger("InnerIFrameDriver");
+    }
 
     // TODO: Issue-2109 Implement detach container api or put appropriate comment.
     public get resolvedUrl(): IResolvedUrl {
+        const event = { category: "generic", eventName: "not implemented" };
+        this.outerProxy.logger.send(event);
+        this.logger.send(event);
```
Comment (ChumpChief) [inline packages/drivers/iframe-driver/src/innerDocumentService.ts:53]: Is this intentionally sending twice?  I think I'm also not really understanding in what situations this should be logging on its own logger vs. trying to access the outerProxy's logger.  It feels a little strange to me to be directly accessing another object's logger.

Comment (heliocliu) [reply packages/drivers/iframe-driver/src/innerDocumentService.ts:53]: it's sending twice so both its own and the outer proxy logger get used.  it's definitely not "right" but more for the example atm.

i don't think it's _that_ crazy when the other object is the object we're proxying.  it seems like we might want to log from the outer driver when the inner driver doesn't need its own logger (like in our first party case) and log from the inner when it does (custom 3rd part logger)?  the main thing was that both options should be available if either can ever a better call than the other, but if we think one is just always going to be better then removing the option seems fine

File: packages/drivers/iframe-driver/src/debug.ts [hunk 4]
```diff
@@ -4,7 +4,18 @@
  */
 
 import registerDebug from "debug";
+import { ITelemetryBaseEvent, ITelemetryBaseLogger } from "@fluidframework/common-definitions";
 import { pkgName, pkgVersion } from "./packageVersion";
 
 export const debug = registerDebug("fluid:iframe-socket-storage");
 debug(`Package: ${pkgName} - Version: ${pkgVersion}`);
+
+export class IFrameDebugLogger implements ITelemetryBaseLogger {
```
Comment (anthony-murphy) [inline packages/drivers/iframe-driver/src/debug.ts:13]: we already have a DebugLogger. we should just use that

File: packages/drivers/iframe-driver/src/innerDocumentService.ts (hunk 3 above)
Comment (anthony-murphy) [inline packages/drivers/iframe-driver/src/innerDocumentService.ts:53]: we should have to call two seperate loggers. there should just be a single combined logger

File: packages/drivers/iframe-driver/src/innerDocumentServiceFactory.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/drivers/iframe-driver/src/innerDocumentServiceFactory.ts:32]: or replace with call to logger

---
In reply to: [512853804](https://github.com/microsoft/FluidFramework/pull/4106#discussion_r512853804) [](ancestors = 512853804)

File: packages/drivers/iframe-driver/src/innerDocumentService.ts [hunk 5]
```diff
@@ -22,16 +25,32 @@ export class InnerDocumentService implements IDocumentService {
...
         private readonly outerProxy: ICombinedDriver,
-        public clientId: string) { }
+        public clientId: string,
+        logger?: ITelemetryBaseLogger)
+    {
+        // outerProxy provides a logger as well, but we can create and use our own here
+        // for different functionality
+        this.logger = logger
+            ? ChildLogger.create(logger, "InnerIFrameDriver")
+            : new IFrameDebugLogger("InnerIFrameDriver");
```
Comment (anthony-murphy) [inline packages/drivers/iframe-driver/src/innerDocumentService.ts:46]: i don't think we should create a logger by default. we should take an optional logger, and produce a combined logger with that and the outproxy logger

File: packages/drivers/iframe-driver/src/outerDocumentServiceFactory.ts [hunk 6]
```diff
@@ -69,9 +76,10 @@ export class DocumentServiceFactoryProxy implements IDocumentServiceFactoryProxy
         this.createProxy(frame);
     }
 
-    public async createDocumentService(): Promise<string> {
+    public async createDocumentService(logger?: ITelemetryBaseLogger): Promise<string> {
```
Comment (anthony-murphy) [inline packages/drivers/iframe-driver/src/outerDocumentServiceFactory.ts:79]: >logger [](start = 39, length = 6)

shouldn't this logger just come through the constructor?

File: packages/drivers/iframe-driver/src/debug.ts (hunk 4 above)
Comment (anthony-murphy) [reply packages/drivers/iframe-driver/src/debug.ts:13]: i think this whole file can be deleted

---
In reply to: [513620197](https://github.com/microsoft/FluidFramework/pull/4106#discussion_r513620197) [](ancestors = 513620197)

File: packages/drivers/iframe-driver/src/innerDocumentService.ts (hunk 5 above)
Comment (anthony-murphy) [reply packages/drivers/iframe-driver/src/innerDocumentService.ts:46]: MultiSinkLogger is the combined logger

---
In reply to: [513623391](https://github.com/microsoft/FluidFramework/pull/4106#discussion_r513623391) [](ancestors = 513623391)

File: packages/drivers/iframe-driver/src/debug.ts (hunk 4 above)
Comment (heliocliu) [reply packages/drivers/iframe-driver/src/debug.ts]: the debug itself is still being used in a few places, so going to hold off on removing that for now

**Review by anthony-murphy — APPROVED** [review]
:shipit:

**Pull request merged**